pkg/kubectl/cmd/util/factory_client_access.go

```diff
@@ -596,6 +600,42 @@ func FallbackGeneratorNameIfNecessary(
 	cmdErr io.Writer,
 ) (string, error) {
 	switch generatorName {
+	case DeploymentAppsV1GeneratorName:
+		hasResource, err := HasResource(discoveryClient, appsv1.SchemeGroupVersion.WithResource("deployments"))
+		if err != nil {
+			return "", err
+		}
+		if !hasResource {
+			warning(cmdErr, DeploymentAppsV1GeneratorName, DeploymentAppsV1Beta1GeneratorName)
+			return FallbackGeneratorNameIfNecessary(DeploymentAppsV1Beta1GeneratorName, discoveryClient, cmdErr)
```

Comment:
This essentially checks recursively if the older generator can be used. The order is apps/v1 -> apps/v1beta1 -> extensions/v1beta -> v1 ReplicationController

Reply:
shouldn't you wait to print the warning until you know what version you're going to use?

Reply:
Wouldn't be cleaner to just return the generator name `if hasResource` and use `fallthrough` otherwise, printing the warning when `generatorName != <what we're about to return to user>` just like Jordan suggested?